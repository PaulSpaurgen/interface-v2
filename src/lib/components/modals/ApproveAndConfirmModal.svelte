<script lang="ts">
	import Button from '$lib/atoms/buttons/Button.svelte';
	import Modal from '$lib/atoms/modals/Modal.svelte';
	import Text from '$lib/atoms/texts/Text.svelte';
	import SuccessfulConversionModal from '$lib/page-components/bridge/modals/SuccessfulConversionModal.svelte';
	import { amountPrecision, BigNumberZero } from '$lib/utils/constants/constants';
	import { bigNumberToCommaString, mPondToPond, pondToMPond } from '$lib/utils/conversion';
	import { closeModal, openModal } from '$lib/utils/helpers/commonHelper';
	import type { BigNumber } from 'ethers';
	import LoadingAnimatedPing from '../loading/LoadingAnimatedPing.svelte';

	export let handleApproveClick: () => Promise<void>;
	export let handleConfirmClick: () => Promise<void>;
	export let handleSuccessFinishClick: () => void;
	export let approved: boolean;
	export let rowIndex: number;
	export let approveButtonText = 'APPROVE';
	export let confirmButtonText = 'CONFIRM';
	export let amountConverted: BigNumber = BigNumberZero;
	export let conversionFrom: 'pond' | 'mPond' = 'pond';

	export let modalForApproveConfirm: string;

	$: amountConvertedFrom = bigNumberToCommaString(amountConverted, amountPrecision(conversionFrom));
	$: amountConvertedTo = bigNumberToCommaString(
		conversionFrom === 'pond' ? pondToMPond(amountConverted) : mPondToPond(amountConverted),
		amountPrecision(conversionFrom === 'pond' ? 'mPond' : 'pond')
	);
	$: conversionFromText = conversionFrom === 'pond' ? 'POND' : 'MPond';
	$: conversionToText = conversionFrom === 'pond' ? 'MPond' : 'POND';
	$: modalForSuccessConversion = `success-conversion-modal-${rowIndex}`;
	$: console.log(modalForSuccessConversion);
	let approveLoading: boolean;
	let confirmLoading: boolean;

	const approveClick = async () => {
		try {
			approveLoading = true;
			await handleApproveClick();
			approveLoading = false;
		} catch (e) {
			approveLoading = false;
			throw e;
		}
	};

	const confirmClick = async () => {
		try {
			confirmLoading = true;
			await handleConfirmClick();
			handleSuccessFinishClick();
			closeModal(modalForApproveConfirm);
			openModal(modalForSuccessConversion);
		} catch (e) {
			throw e;
		} finally {
			confirmLoading = false;
		}
	};
	// const modalWidth = 'max-w-[500px]';

	const styles = {
		baseText: 'grow text-left  font-normal',
		text: 'text-grey-500 font-[15px]',
		textBold: 'font-semibold text-black',
		textDisabled: 'font-semibold text-grey-300'
	};
</script>

<Modal modalFor={modalForApproveConfirm}>
	<svelte:fragment slot="title">
		{!approved ? 'Approve Transaction' : 'Confirm Transaction'}
	</svelte:fragment>
	<svelte:fragment slot="content">
		<div class="flex gap-5 h-[50px]">
			<div class="flex flex-col items-center">
				{#if approved}
					<img src="/images/vectorcheck.svg" alt="Copy" width="20px" height="20px" />
				{:else}
					<LoadingAnimatedPing loading={approveLoading}>
						<Text variant="small" styleClass="font-semibold" text={'1'} />
					</LoadingAnimatedPing>
				{/if}
				{#if approved}
					<div class="h-full w-[0.1px] bg-grey-400" />
				{/if}
			</div>
			<div class={`${styles.text} ${styles.baseText}`}>
				<span>{'Approve'}</span>
				<span class={styles.textBold}>
					{`${amountConvertedFrom} ${conversionFromText}`}
				</span>
				<span>{'for conversion'}</span>
			</div>
		</div>
		<div class={`flex gap-5`}>
			<LoadingAnimatedPing loading={confirmLoading}>
				<Text variant="small" styleClass="font-semibold" text={'2'} />
			</LoadingAnimatedPing>
			<div class={`${approved ? styles.text : 'text-grey-300'} ${styles.baseText}`}>
				<span>{'Convert'}</span>
				<span class={approved ? styles.textBold : styles.textDisabled}>
					{`${amountConvertedFrom} ${conversionFromText}`}
				</span>
				<span>{'to'}</span>
				<span class={approved ? styles.textBold : styles.textDisabled}>
					{`${amountConvertedTo} ${conversionToText}`}
				</span>
			</div>
		</div>
	</svelte:fragment>
	<svelte:fragment slot="actionButtons">
		{#if !approved}
			<Button variant="filled" size="large" loading={approveLoading} onclick={approveClick}>
				{approveButtonText}
			</Button>
		{:else}
			<Button
				size="large"
				variant="filled"
				styleClass="w-full"
				onclick={confirmClick}
				loading={confirmLoading}
			>
				{confirmButtonText}
			</Button>
		{/if}
	</svelte:fragment>
</Modal>
<SuccessfulConversionModal
	modalFor={modalForSuccessConversion}
	{amountConverted}
	{conversionFrom}
/>
