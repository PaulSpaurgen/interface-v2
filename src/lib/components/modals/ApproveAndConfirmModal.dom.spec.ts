import { describe, it, expect, vi } from 'vitest';
import { render, fireEvent } from '@testing-library/svelte';
import ApproveAndConfirmModal from './ApproveAndConfirmModal.svelte';
import { tick } from 'svelte';


describe('Component', () => {
    it(`render's the component properly`, () => {
        const handleApproveClick = vi.fn().mockImplementation(() => Promise.resolve());
        const handleConfirmClick = vi.fn().mockImplementation(() => Promise.resolve());
        const { container } = render(ApproveAndConfirmModal, {
            handleApproveClick,
            approved: false,
            handleConfirmClick,
            rowIndex: 10,
            modalForApproveConfirm: 'test-id-some-id',
            handleSuccessFinishClick: () => {
            },
            conversionFrom: "mPond",
            amountConverted: 123999999999n,
            confirmButtonText: "CONVERT",
        });
        expect(container).toMatchSnapshot();
    })
    it('should call handleApproveClick when approve button is clicked and render relevant elements', async () => {
        const handleApproveClick = vi.fn().mockImplementation(() => Promise.resolve());
        const handleConfirmClick = vi.fn().mockImplementation(() => Promise.resolve());
        const { getByText } = render(ApproveAndConfirmModal, {
            handleApproveClick,
            approved: false,
            handleConfirmClick,
            rowIndex: 10,
            modalForApproveConfirm: 'test-id-some-id',
            handleSuccessFinishClick: () => {
            },
            conversionFrom: "mPond",
            amountConverted: 123999999999n,
            confirmButtonText: "CONVERT",
        });
        expect(getByText('Approve Transaction')).toBeTruthy();
        await fireEvent.click(getByText('APPROVE'));
        await tick();
        expect(handleApproveClick).toHaveBeenCalledOnce();
    });

    it('should call handleConfirmClick when confirm button is clicked and render relevant elements', async () => {
        const handleConfirmClick = vi.fn().mockImplementation(() => Promise.resolve());
        const handleSuccessFinishClick = vi.fn();
        const handleApproveClick = vi.fn().mockImplementation(() => Promise.resolve());

        const { getByText } = render(ApproveAndConfirmModal, {
            handleApproveClick,
            approved: true,
            handleConfirmClick,
            rowIndex: 10,
            modalForApproveConfirm: 'test-id-some-id',
            handleSuccessFinishClick,
            conversionFrom: "mPond",
            amountConverted: 123999999999n,
            confirmButtonText: "CONVERT",
        });
        // implementation if using html
        // const { getByText } = render(html`<${ApproveAndConfirmModal} 
        //     handleApproveClick=${handleApproveClick}
        //     approved=${true}
        //     handleConfirmClick=${handleConfirmClick}
        //     rowIndex=${10}
        //     modalForApproveConfirm=${'test-id-some-id'}
        //     handleSuccessFinishClick=${handleSuccessFinishClick}
        //     conversionFrom=${"mPond"}
        //     amountConverted=${123999999999n}
        //     confirmButtonText=${"CONVERT"}
        // >

        //     </${ApproveAndConfirmModal}>
        //   `)

        expect(getByText('Confirm Transaction')).toBeTruthy();
        await fireEvent.click(getByText('CONVERT'));

        expect(handleConfirmClick).toHaveBeenCalled();
        await tick()
        expect(handleSuccessFinishClick).toHaveBeenCalled();
    });

});