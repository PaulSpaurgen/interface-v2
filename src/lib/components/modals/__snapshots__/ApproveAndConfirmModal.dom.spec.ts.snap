// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ApproveAndConfirmModal > render's the ApproveAndConfirmModal properly 1`] = `
<body>
  <onboard-v2
    style="all: initial;"
  />
  <style>
    
    
@font-face {
  font-family: 'Inter';
  font-style:  normal;
  font-weight: 300 600;
  font-display: swap;
  src: url("https://rsms.me/inter/font-files/Inter-roman.var.woff2?v=3.19") format("woff2");
}

  
  </style>
  <div>
    <input
      class="modal-toggle s-GZwqZrPaBy6h"
      id="test-id-some-id"
      tabindex="-1"
      type="checkbox"
    />
     
    <div
      class="modal s-GZwqZrPaBy6h"
      data-testid="modal"
    >
      <div
        class="w-11/12 sm:w-3/4 sm:max-w-[607px] modal-box rounded-lg bg-base-100 px-0 py-0 shadow-none s-GZwqZrPaBy6h"
      >
        <div
          class="modal-header sticky top-0 z-10 flex items-center bg-white px-6 pb-4 pt-8 s-GZwqZrPaBy6h"
        >
          <div
            class="flex w-full flex-col s-GZwqZrPaBy6h"
          >
             
            <div
              class="text-left text-[15px] font-medium text-[#0a0e3099] s-GZwqZrPaBy6h"
            />
             
            <div
              class="text-left text-2xl font-bold s-GZwqZrPaBy6h"
            >
              Approve Transaction
            </div>
             
            <div
              class="mt-1 text-left text-[15px] font-medium text-[#767676] s-GZwqZrPaBy6h"
            />
          </div>
           
          <button
            class="flex h-12 min-w-[3rem] cursor-pointer items-center justify-center self-start text-sm font-bold text-gray-300 s-GZwqZrPaBy6h"
            data-testid="modal-close-button"
          >
            <img
              alt="Close"
              class="s-GZwqZrPaBy6h"
              src="/images/close-circle-icon.svg"
            />
          </button>
        </div>
         
        <div
          class="modal-body  px-6 pb-4 pt-2 s-GZwqZrPaBy6h"
        >
          <div
            class="flex h-[50px] gap-5"
          >
            <div
              class="flex flex-col items-center"
            >
              <span
                class="relative flex justify-center rounded-3xl text-white h-5 w-5"
              >
                <span
                  class="absolute inline-flex h-full w-full animate-ping rounded-full bg-orange-400 opacity-75"
                />
                 
                <span
                  class="relative inline-flex justify-center rounded-full bg-orange-400 text-center text-sm h-5 w-5"
                >
                  <div
                    class="font-semibold text-sm font-normal"
                    data-testid="text"
                  >
                    1
                  </div>
                  <!--&lt;Text&gt;-->
                </span>
              </span>
              
              <!--&lt;LoadingAnimationModal&gt;-->
               
            </div>
             
            <div
              class="grow text-left font-normal text-grey-500"
            >
              Approve
				
              <span
                class="font-semibold text-black"
              >
                0.000000
                 
                MPond
              </span>
              
				for conversion
            </div>
          </div>
           
          <div
            class="flex gap-5"
          >
            <div
              class="flex justify-center rounded-3xl bg-grey-400 text-white h-5 w-5"
            >
              <div
                class="font-semibold text-sm font-normal"
                data-testid="text"
              >
                2
              </div>
              <!--&lt;Text&gt;-->
            </div>
            
            <!--&lt;LoadingAnimationModal&gt;-->
             
            <div
              class="grow text-left font-normal text-grey-300"
            >
              Convert
				
              <span
                class="font-semibold text-grey-300"
              >
                0.000000
                 
                MPond
              </span>
              
				to
				
              <span
                class="font-semibold text-grey-300"
              >
                0.12
                 
                POND
              </span>
            </div>
          </div>
        </div>
         
        <div
          class="modal-footer sticky bottom-0 bg-white mb-6 mt-4 px-6 s-GZwqZrPaBy6h"
        >
          <button
            class="w-full h-14 text-base font-semibold btn btn-theme btn-sm items-center flex justify-center disabled:cursor-not-allowed btn-primary disabled:bg-primary disabled:bg-opacity-60 disabled:text-white gap-1.5"
          >
             
            APPROVE
          </button>
          <!--&lt;Button&gt;-->
          
        </div>
      </div>
    </div>
    <!--&lt;Modal&gt;-->
     
    <input
      class="modal-toggle s-GZwqZrPaBy6h"
      id="success-conversion-modal-10"
      tabindex="-1"
      type="checkbox"
    />
     
    <div
      class="modal s-GZwqZrPaBy6h"
      data-testid="modal"
    >
      <div
        class="w-11/12 sm:w-3/4 sm:max-w-[607px] modal-box rounded-lg bg-base-100 px-0 py-0 shadow-none s-GZwqZrPaBy6h"
      >
        <div
          class="modal-header sticky top-0 z-10 flex items-center bg-white px-6 pb-4 pt-8 s-GZwqZrPaBy6h"
        >
          <div
            class="flex w-full flex-col s-GZwqZrPaBy6h"
          >
            <div
              class="mb-6 s-GZwqZrPaBy6h"
            >
              <img
                alt=""
                slot="icon"
                src="/images/shield.svg"
                width="38px"
              />
            </div>
             
            <div
              class="text-left text-[15px] font-medium text-[#0a0e3099] s-GZwqZrPaBy6h"
            />
             
            <div
              class="text-left text-2xl font-bold s-GZwqZrPaBy6h"
            >
              Conversion Successful
            </div>
             
            <div
              class="mt-1 text-left text-[15px] font-medium text-[#767676] s-GZwqZrPaBy6h"
            />
          </div>
           
          <button
            class="flex h-12 min-w-[3rem] cursor-pointer items-center justify-center self-start text-sm font-bold text-gray-300 s-GZwqZrPaBy6h"
            data-testid="modal-close-button"
          >
            <img
              alt="Close"
              class="s-GZwqZrPaBy6h"
              src="/images/close-circle-icon.svg"
            />
          </button>
        </div>
         
        <div
          class="modal-body  px-6 pb-4 pt-2 s-GZwqZrPaBy6h"
        >
          <div
            class="text-left text-base font-medium text-gray-600"
          >
            <div>
              You have converted
            </div>
             
            <div>
              <span
                class="font-bold text-black"
              >
                0.000000
                 
                MPOND
              </span>
              
				to
				
              <span
                class="font-bold text-black"
              >
                0.12
                 
                POND
              </span>
            </div>
             
            <div
              class="h-[1px] w-full bg-black/[0.1] my-6"
              data-testid="divider"
            />
            <!--&lt;Divider&gt;-->
             
            <div>
              Updated Wallet Balance
            </div>
             
            <span
              class="font-bold text-black"
            >
              0.00
               POND
			
            </span>
            |
			
            <span
              class="font-bold text-black"
            >
              0.00
               MPOND
            </span>
          </div>
        </div>
         
        <div
          class="modal-footer sticky bottom-0 bg-white mb-6 mt-4 px-6 s-GZwqZrPaBy6h"
        >
          <button
            class="w-full h-14 text-base font-semibold btn btn-theme btn-sm items-center flex justify-center disabled:cursor-not-allowed btn-primary disabled:bg-primary disabled:bg-opacity-60 disabled:text-white gap-1.5"
          >
             
            FINISH
          </button>
          <!--&lt;Button&gt;-->
        </div>
      </div>
    </div>
    <!--&lt;Modal&gt;-->
    <!--&lt;SuccessfulConversionModal&gt;-->
    <!--&lt;ApproveAndConfirmModal&gt;-->
  </div>
</body>
`;

exports[`Component > render's the component properly 1`] = `
<body>
  <onboard-v2
    style="all: initial;"
  />
  <style>
    
    
@font-face {
  font-family: 'Inter';
  font-style:  normal;
  font-weight: 300 600;
  font-display: swap;
  src: url("https://rsms.me/inter/font-files/Inter-roman.var.woff2?v=3.19") format("woff2");
}

  
  </style>
  <div>
    <input
      class="modal-toggle s-GZwqZrPaBy6h"
      id="test-id-some-id"
      tabindex="-1"
      type="checkbox"
    />
     
    <div
      class="modal s-GZwqZrPaBy6h"
      data-testid="modal"
    >
      <div
        class="w-11/12 sm:w-3/4 sm:max-w-[607px] modal-box rounded-lg bg-base-100 px-0 py-0 shadow-none s-GZwqZrPaBy6h"
      >
        <div
          class="modal-header sticky top-0 z-10 flex items-center bg-white px-6 pb-4 pt-8 s-GZwqZrPaBy6h"
        >
          <div
            class="flex w-full flex-col s-GZwqZrPaBy6h"
          >
             
            <div
              class="text-left text-[15px] font-medium text-[#0a0e3099] s-GZwqZrPaBy6h"
            />
             
            <div
              class="text-left text-2xl font-bold s-GZwqZrPaBy6h"
            >
              Approve Transaction
            </div>
             
            <div
              class="mt-1 text-left text-[15px] font-medium text-[#767676] s-GZwqZrPaBy6h"
            />
          </div>
           
          <button
            class="flex h-12 min-w-[3rem] cursor-pointer items-center justify-center self-start text-sm font-bold text-gray-300 s-GZwqZrPaBy6h"
            data-testid="modal-close-button"
          >
            <img
              alt="Close"
              class="s-GZwqZrPaBy6h"
              src="/images/close-circle-icon.svg"
            />
          </button>
        </div>
         
        <div
          class="modal-body  px-6 pb-4 pt-2 s-GZwqZrPaBy6h"
        >
          <div
            class="flex h-[50px] gap-5"
          >
            <div
              class="flex flex-col items-center"
            >
              <span
                class="relative flex justify-center rounded-3xl text-white h-5 w-5"
              >
                <span
                  class="absolute inline-flex h-full w-full animate-ping rounded-full bg-orange-400 opacity-75"
                />
                 
                <span
                  class="relative inline-flex justify-center rounded-full bg-orange-400 text-center text-sm h-5 w-5"
                >
                  <div
                    class="font-semibold text-sm font-normal"
                    data-testid="text"
                  >
                    1
                  </div>
                  <!--&lt;Text&gt;-->
                </span>
              </span>
              
              <!--&lt;LoadingAnimationModal&gt;-->
               
            </div>
             
            <div
              class="grow text-left font-normal text-grey-500"
            >
              Approve
				
              <span
                class="font-semibold text-black"
              >
                0.000000
                 
                MPond
              </span>
              
				for conversion
            </div>
          </div>
           
          <div
            class="flex gap-5"
          >
            <div
              class="flex justify-center rounded-3xl bg-grey-400 text-white h-5 w-5"
            >
              <div
                class="font-semibold text-sm font-normal"
                data-testid="text"
              >
                2
              </div>
              <!--&lt;Text&gt;-->
            </div>
            
            <!--&lt;LoadingAnimationModal&gt;-->
             
            <div
              class="grow text-left font-normal text-grey-300"
            >
              Convert
				
              <span
                class="font-semibold text-grey-300"
              >
                0.000000
                 
                MPond
              </span>
              
				to
				
              <span
                class="font-semibold text-grey-300"
              >
                0.12
                 
                POND
              </span>
            </div>
          </div>
        </div>
         
        <div
          class="modal-footer sticky bottom-0 bg-white mb-6 mt-4 px-6 s-GZwqZrPaBy6h"
        >
          <button
            class="w-full h-14 text-base font-semibold btn btn-theme btn-sm items-center flex justify-center disabled:cursor-not-allowed btn-primary disabled:bg-primary disabled:bg-opacity-60 disabled:text-white gap-1.5"
          >
             
            APPROVE
          </button>
          <!--&lt;Button&gt;-->
          
        </div>
      </div>
    </div>
    <!--&lt;Modal&gt;-->
     
    <input
      class="modal-toggle s-GZwqZrPaBy6h"
      id="success-conversion-modal-10"
      tabindex="-1"
      type="checkbox"
    />
     
    <div
      class="modal s-GZwqZrPaBy6h"
      data-testid="modal"
    >
      <div
        class="w-11/12 sm:w-3/4 sm:max-w-[607px] modal-box rounded-lg bg-base-100 px-0 py-0 shadow-none s-GZwqZrPaBy6h"
      >
        <div
          class="modal-header sticky top-0 z-10 flex items-center bg-white px-6 pb-4 pt-8 s-GZwqZrPaBy6h"
        >
          <div
            class="flex w-full flex-col s-GZwqZrPaBy6h"
          >
            <div
              class="mb-6 s-GZwqZrPaBy6h"
            >
              <img
                alt=""
                slot="icon"
                src="/images/shield.svg"
                width="38px"
              />
            </div>
             
            <div
              class="text-left text-[15px] font-medium text-[#0a0e3099] s-GZwqZrPaBy6h"
            />
             
            <div
              class="text-left text-2xl font-bold s-GZwqZrPaBy6h"
            >
              Conversion Successful
            </div>
             
            <div
              class="mt-1 text-left text-[15px] font-medium text-[#767676] s-GZwqZrPaBy6h"
            />
          </div>
           
          <button
            class="flex h-12 min-w-[3rem] cursor-pointer items-center justify-center self-start text-sm font-bold text-gray-300 s-GZwqZrPaBy6h"
            data-testid="modal-close-button"
          >
            <img
              alt="Close"
              class="s-GZwqZrPaBy6h"
              src="/images/close-circle-icon.svg"
            />
          </button>
        </div>
         
        <div
          class="modal-body  px-6 pb-4 pt-2 s-GZwqZrPaBy6h"
        >
          <div
            class="text-left text-base font-medium text-gray-600"
          >
            <div>
              You have converted
            </div>
             
            <div>
              <span
                class="font-bold text-black"
              >
                0.000000
                 
                MPOND
              </span>
              
				to
				
              <span
                class="font-bold text-black"
              >
                0.12
                 
                POND
              </span>
            </div>
             
            <div
              class="h-[1px] w-full bg-black/[0.1] my-6"
              data-testid="divider"
            />
            <!--&lt;Divider&gt;-->
             
            <div>
              Updated Wallet Balance
            </div>
             
            <span
              class="font-bold text-black"
            >
              0.00
               POND
			
            </span>
            |
			
            <span
              class="font-bold text-black"
            >
              0.00
               MPOND
            </span>
          </div>
        </div>
         
        <div
          class="modal-footer sticky bottom-0 bg-white mb-6 mt-4 px-6 s-GZwqZrPaBy6h"
        >
          <button
            class="w-full h-14 text-base font-semibold btn btn-theme btn-sm items-center flex justify-center disabled:cursor-not-allowed btn-primary disabled:bg-primary disabled:bg-opacity-60 disabled:text-white gap-1.5"
          >
             
            FINISH
          </button>
          <!--&lt;Button&gt;-->
        </div>
      </div>
    </div>
    <!--&lt;Modal&gt;-->
    <!--&lt;SuccessfulConversionModal&gt;-->
    <!--&lt;ApproveAndConfirmModal&gt;-->
  </div>
</body>
`;
